cmd/engine/executor.go

```diff
@@ -0,0 +1,1053 @@
+		defer clean()
+	}
+
+	configPath := filepath.Join(bundle, "config.json")
+	f, err := os.Create(configPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	opts := []containerdoci.SpecOpts{oci.WithUIDGID(uid, gid, sgids)}
+
+	if process.Meta.ReadonlyRootFS {
+		opts = append(opts, containerdoci.WithRootFSReadonly())
+	}
+
+	identity := idtools.Identity{
+		UID: int(uid),
+		GID: int(gid),
+	}
+	if w.idmap != nil {
+		identity, err = w.idmap.ToHost(identity)
+		if err != nil {
+			return err
+		}
+	}
+
+	spec, cleanup, err := oci.GenerateSpec(
+		ctx,
+		process.Meta,
+		mounts,
+		id,
+		resolvConf,
+		hostsFile,
+		namespace,
+		w.cgroupParent,
+		w.processMode,
+		w.idmap,
+		w.apparmorProfile,
+		w.selinux,
+		w.tracingSocket,
+		opts...,
+	)
+	if err != nil {
+		return err
+	}
+	defer cleanup()
+
+	spec.Root.Path = rootFSPath
+	if root.Readonly {
+		spec.Root.Readonly = true
+	}
+
+	newp, err := fs.RootPath(rootFSPath, process.Meta.Cwd)
+	if err != nil {
+		return errors.Wrapf(err, "working dir %s points to invalid target", newp)
+	}
+	if _, err := os.Stat(newp); err != nil {
+		if err := idtools.MkdirAllAndChown(newp, 0o755, identity); err != nil {
+			return errors.Wrapf(err, "failed to create working directory %s", newp)
+		}
+	}
+
+	spec.Process.Terminal = process.Meta.Tty
+	spec.Process.OOMScoreAdj = w.oomScoreAdj
+
+	if err := json.NewEncoder(f).Encode(spec); err != nil {
+		return err
+	}
+	f.Close()
+
+	bklog.G(ctx).Debugf("> creating %s %v", id, process.Meta.Args)
+	defer bklog.G(ctx).Debugf("> container done %s %v", id, process.Meta.Args)
+
+	if installCACerts {
+		caInstaller, err := cacerts.NewInstaller(ctx, spec, func(ctx context.Context, args ...string) error {
+			id := randid.NewID()
+			meta := process.Meta
+			meta.Args = args
+			meta.User = "0:0"
+			meta.Cwd = "/"
+			meta.Tty = false
+			output := new(bytes.Buffer)
+			process := executor.ProcessInfo{
+				Stdout: nopCloser{output},
+				Stderr: nopCloser{output},
+				Meta:   meta,
+			}
+			started := make(chan struct{}, 1)
+			installCACerts := false
+			err := w.run(
+				ctx,
+				id,
+				root,
+				mounts,
+				rootFSPath,
+				process,
+				namespace,
+				started,
+				installCACerts,
+			)
+			return err
+		})
+		if err == nil {
+			err = caInstaller.Install(ctx)
+			if errors.As(err, new(cacerts.CleanupErr)) {
+				// if install failed and cleanup failed too, we have no choice but to fail this exec; otherwise we're
+				// leaving the container in some weird half state
+				return fmt.Errorf("failed to install cacerts: %w", err)
+			}
+			// if install failed but we were able to cleanup, then we should log it but don't need to fail the exec
+			if err != nil {
+				bklog.G(ctx).Errorf("failed to install cacerts but successfully cleaned up, continuing without CA certs: %v", err)
+			} else {
+				defer func() {
+					if err := caInstaller.Uninstall(ctx); err != nil {
+						bklog.G(ctx).Errorf("failed to uninstall cacerts: %v", err)
```

Comment:
Paranoia brain: if uninstall fails, should we fail the whole thing for the same reason as the cleanup failure?

Worst case I guess is that we "leak" a CA cert into the container data? But that's not really a sensitive value, just a little odd to leave it behind

Reply:
Yeah I wasn't totally sure here, but in retrospect it's safer to start with an error and make non fatal than vice versa